### PR TITLE
Still trying to fix notification caching when un/subscribing

### PR DESCRIPTION
**Pull request by jaw6**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [ ] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

After deploying #19555, there are reports of buttons being "sticky" between refreshes. It turns out that the notification fragments are cached and we aren't clearing them. In #19739, I was trying to leave cache-keys as-is and burst them during subscription changes, but it turns out that this is too slow in production. In this PR, I am instead altering the cache-keys to insert a "subscription status indicator" — in this case, I am doing a `sum` of all the current user's subscription IDs. (This is memoized, so the math should only happen once, and apply to all the user's notification fragments.) In this way, _any_ change to the user's subscriptions (destroying or creating a subscription) will change the "indicator" and the user's next view of `/notifications` will have a cache miss.

## Related Tickets & Documents

- Related Issue #19350 

## QA Instructions, Screenshots, Recordings

`rails dev:cache` will toggle caching behavior in developer local environment.

## Added/updated tests?

- [ ] Yes
- [x] No, and this is why: we don't have a rigorous process for testing views and helpers
- [ ] I need help with writing tests

## [optional] Are there any post deployment tasks we need to perform?

An unpleasant side-effect here is that this PR will cause all users' `/notifications` views to miss cache. (These keys incorporate the user's last reaction and last comment IDs already, so I _think_ we're okay with these being fairly volatile, but I thought I should mention it anyway!)

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

File: app/helpers/notifications_helper.rb [hunk 1]
```diff
@@ -50,4 +50,11 @@ def mod_article_user(data)
     article_username = data["article"]["path"].split("/")[1]
     { "name" => article_username, "path" => "/#{article_username}" }
   end
+
+  # This is used in the notification view's cache key, so that cached
+  # notification fragments can quickly be burst whenever a user
+  # adjusts their subscriptions to comment notifications
+  def subscription_status_indicator
+    @subscription_status_indicator ||= current_user.notification_subscriptions.sum(:id)
```
Comment (Ridhwana) [inline app/helpers/notifications_helper.rb:58]: Question: Why do we use sum(:id) here instead of the count?

**Review by Ridhwana — APPROVED** [review]
This looks okay to me, just has a small question.

File: app/helpers/notifications_helper.rb (hunk 1 above)
Comment (jaw6) [reply app/helpers/notifications_helper.rb:58]: So, what I saw was that it's possible for a user to unsubscribe and re-subscribe (or subscribe to thread A and unsubscribe to thread B). In those sorts of scenarios, the count would remain the same, but we would _want_ the cache miss. `sum` is better able to handle scenarios where any `-` is balanced by a `+`. I _think_ we have elsewhere tended to use `last_updated_at`, but that doesn't work as well here, where `unsubscribe` **destroys** the object (we don't want to soft-delete `subscriptions`, there are too many records for that to work).

Comment (Ridhwana) [reply app/helpers/notifications_helper.rb:58]: This makes thanks, thank you for explaining.

**Review by benhalpern — APPROVED** [review]
This looks good to me.

I imagine in the event of the memoized `subscription_status_indicator` not turning out to be efficient enough in the future, this could be swapped for a "last action at" attribute like we have elsewhere. Just making note of that, maybe we actually want to get _further_ away from over-use of those event caches.

File: app/helpers/notifications_helper.rb (hunk 1 above)
Comment (maestromac) [reply app/helpers/notifications_helper.rb:58]: I _think_ you can also use cache key here. It's combination of the query digest + result count + last_updated_at on the last titem.
`current_user.notification_subscriptions.cache_key_with_version`
ref https://www.bigbinary.com/blog/activerecord-relation-cache-key

**Pull request merged**